### PR TITLE
hash_transform snark with int128

### PR DESCRIPTION
**Pull request by josojo**
Finally, it works.

Thanks again for your help.

### CODE REVIEW

File: pepper/Dockerfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ MAINTAINER Felix Leupold <felix@gnosis.pm>
 ENV PEPPER /home/docker/pequin/
 
 USER root
-RUN apt-get install -y inotify-tools libgtest-dev time vim tmux
+RUN apt-get install -y inotify-tools libgtest-dev time tmux
```
Comment (fleupold) [inline pepper/Dockerfile]: Can we find a solution for this while keeping vim installed? Also not really related to this PR so could be its own commit.

File: pepper/apps/hash_transform.c [hunk 2]
```diff
@@ -7,6 +7,8 @@ struct PedersenHash { uint32_t values[2]; }; /* [x,y] */
 void compute(struct In *input, struct Out *output){
     // Read private input
     struct Private pInput = readPrivateInput();
+    //printf("LHS of hash input %Zd and RHS of hash input %Zd ", input->shaHashL, input->shaHashR);
```
Comment (fleupold) [inline pepper/apps/hash_transform.c]: Remove please

File: pepper/prover_verifier_shared/transform.inputs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+062174266130574243119709474530137211798
```
Comment (fleupold) [inline pepper/prover_verifier_shared/transform.inputs]: Let's not commit this file

File: pepper/order_transform/generate.py [hunk 4]
```diff
@@ -16,11 +16,12 @@ def main(numOrders):
 		x = pedersenHash.x if pedersenHash else 0
 		pedersenHash = hash_pedersen(x, order)
 		shaHash = hash_sha(shaHash, order)
-
+		
 	# output sha hash
 	# output pedersen hash
 	print(pedersenHash)
 	print(shaHash)
```
Comment (fleupold) [inline pepper/order_transform/generate.py]: I guess this line is now obsolete?

File: pepper/order_transform/generate.py [hunk 5]
```diff
@@ -30,9 +31,12 @@ def hash_pedersen(left, right):
 	H = Point.from_x(FQ(16540640123574156134436876038791482806971768689494387082833631921987005038935))
 	return G * left + H * right
 
+def transform_int128(hash):
```
Comment (fleupold) [inline pepper/order_transform/generate.py:35]: Can you add a comment here describing what the method does? It's using some fancy python logic. It might also be slightly more explicit to return a pair (left, right) and print that instead.

**Review by fleupold — COMMENTED** [review]
Looks really good. Some small comments here and there. But overall, this is a great improvement and will give us full entropy on the sha hash.

File: pepper/apps/hash_transform.c [hunk 6]
```diff
@@ -1,6 +1,10 @@
 #include "hash_transform.h"
 #include "dex_common.h"
 
+#ifdef TEST
+typedef long int128;
```
Comment (fleupold) [inline pepper/apps/hash_transform.c]: This should only be required once (in dex_common.h). Can we remove it here?

File: pepper/apps/hash_transform.h [hunk 7]
```diff
@@ -1,11 +1,16 @@
 #include <stdint.h>
 
 #ifndef TEST
-#define ORDERS 725
+#define ORDERS 2
 #endif
 
+#ifdef TEST
+typedef long int128;
```
Comment (fleupold) [inline pepper/apps/hash_transform.h]: ditto.

File: pepper/prover_verifier_shared/transform.inputs (hunk 3 above)
Comment (fleupold) [reply pepper/prover_verifier_shared/transform.inputs]: Please remove this file

File: pepper/test/apps/util.h [hunk 8]
```diff
@@ -61,6 +61,6 @@ void exo_compute(uint32_t** input, uint32_t* length, void* output, uint32_t exo)
     }
 }
 
-void assert_zero(uint32_t v) {
+void assert_zero(unsigned int v) {
```
Comment (fleupold) [inline pepper/test/apps/util.h]: why that change?

**Review by fleupold — COMMENTED** [review]
A lot of files have their permissions changed from 100644 to 100755 (meaning they are now executable, which I don't believe they should). Can we revert this please?

Some more small comments inline.

File: pepper/order_transform/generate.py [hunk 9]
```diff
@@ -30,9 +30,14 @@ def hash_pedersen(left, right):
 	H = Point.from_x(FQ(16540640123574156134436876038791482806971768689494387082833631921987005038935))
 	return G * left + H * right
 
+#transform_int128 transform a 256 bit value into two 128 bit values by cutting the hash 
+# into two substrings and representing these substrings as ints with 39 digits 
+def transform_int128(hash):
+	s = hex(hash)
+	return [str(int(s[:34],0)).zfill(39),str(int(s[34:],16)).zfill(39)]
```
Comment (fleupold) [inline pepper/order_transform/generate.py:37]: Just a small question (no need to change), but why do we have to zero fill the ints?

Comment (josojo) [reply pepper/order_transform/generate.py:37]: there is no need. 
Previously, I had the numbers printed as one big number and then split the numbers by hand. For this procedure, it was handy, but it is not needed for sure

File: pepper/apps/hash_transform.h (hunk 7 above)
Comment (fleupold) [reply pepper/apps/hash_transform.h]: I think you might have to import common.h here for it to pick up the uint typedef from there

**Comment by josojo** [comment]
I needed to move the struct Private variable into the dex_common, otherwise I would get this error during the e2e test:

BUILD SUCCESSFUL
Total time: 0 seconds
make[2]: Leaving directory `/home/docker/pequin/compiler/frontend'
WARNING: --cstdarithtruncate is disabled, so type errors will warn and arithmetic is not ANSI C compliant
Compiling hash_transform.c
Expanding circuit to file hash_transform.c.ZAATAR.circuit
Exception in thread "main" java.lang.RuntimeException: No function definition for readPrivateInput
	at ccomp.parser_hw.CCompiler.genericFunctionCall(CCompiler.java:1012)
	at ccomp.parser_hw.CCompiler.getExpr(CCompiler.java:953)
	at ccomp.parser_hw.CCompiler.addInitializerStatement(CCompiler.java:351)
	at ccomp.parser_hw.CCompiler.expandStatement(CCompiler.java:619)
	at ccomp.parser_hw.CCompiler.access$1100(CCompiler.java:109)
	at ccomp.parser_hw.CCompiler$CMainStatement.toAssignmentStatements(CCompiler.java:486)
	at SFE.Compiler.BlockStatement.toAssignmentStatements(BlockStatement.java:111)
	at SFE.Compiler.FunctionBody.toAssignmentStatements(FunctionBody.java:31)
	at zcc.ZCC.compile(ZCC.java:260)
	at zcc.ZCC.main(ZCC.java:127)

I don't understand this
Do you have any clue why? @fleupold

**Comment by fleupold** [comment]
It's because the method `readPrivateInput` uses the Private struct, and if dex-common.h is included before Private is defined it cannot find the type. However, we cannot define Private in dex-common.h, because it won't be common for all circuits.
The main verification circuit will have a different private input struct than the hash transformation circuit.

Can you try forward declaring the struct in dex-common.h like here: https://stackoverflow.com/questions/9999512/forward-declaration-of-a-struct-in-c?
